### PR TITLE
Support auth message both client side and server side.

### PR DESCRIPTION
**Pull request by qianwj**
Motivation:

At the moment of writing, the vertx-mqtt supports all MQTT version 5.0 features except AUTH message which is yet to be implemented. Now I used it to develop my mqtt broker. So, I try to finished this.

Conformance:

Your commits should be signed and you should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md
Please also make sure you adhere to the code style guidelines: https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines

### CODE REVIEW

**Comment by vietj** [comment]
can you sign the eclipse contributor agreement @qianwj ?

**Comment by qianwj** [comment]
> 你能签署 Eclipse 贡献者协议吗 @qianwj ?

**Comment by qianwj** [comment]
I just signed up.

File: src/main/java/io/vertx/mqtt/messages/MqttAuthenticationExchangeMessage.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+public interface MqttAuthenticationExchangeMessage {
+
+  /**
+   * Create a concrete instance of a Vert.x auth message
+   *
+   * @param reasonCode authenticate reason code
+   * @param properties mqtt properties.
+   * @return Vert.x auth message
+   */
+  @GenIgnore
```
Comment (vietj) [inline src/main/java/io/vertx/mqtt/messages/MqttAuthenticationExchangeMessage.java:24]: `@GenIgnore(GenIgnore.PERMITED_TYPES)`

**Comment by vietj** [comment]
The contribution looks great, can you add documentation for this in index.adoc ?

**Comment by qianwj** [comment]
done. [https://github.com/vert-x3/vertx-mqtt/pull/248/commits/a69174152451cc635f7f6a6f432b1ee448a32353](https://github.com/vert-x3/vertx-mqtt/pull/248/commits/a69174152451cc635f7f6a6f432b1ee448a32353)

**Comment by qianwj** [comment]
hi @vietj  , I hope you're doing well. It has been almost a month since I submitted my code changes for review. I understand you might be busy, but I wanted to kindly remind you about my pull [https://github.com/vert-x3/vertx-mqtt/pull/248](https://github.com/vert-x3/vertx-mqtt/pull/248) and request your review.

File: src/main/java/io/vertx/mqtt/MqttClient.java [hunk 2]
```diff
@@ -202,6 +203,23 @@ static MqttClient create(Vertx vertx) {
...
+   * Sets handler which will be called after AUTH packet receiving
+   *
+   * @param authenticationExchangeHandler handler to call
+   * @return current MQTT client instance
+   */
+  @Fluent
+  MqttClient authenticationExchangeHandler(Handler<MqttAuthenticationExchangeMessage> authenticationExchangeHandler);
+
+  /**
+   * Unsubscribe from receiving messages on given list of topic
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/MqttClient.java:216]: I think this comment is wrong, coming from a copy/paste, or?

File: src/main/java/io/vertx/mqtt/MqttEndpoint.java [hunk 3]
```diff
@@ -337,6 +338,15 @@ public interface MqttEndpoint {
   @Fluent
   MqttEndpoint publishCompletionMessageHandler(Handler<MqttPubCompMessage> handler);
 
+  /**
+   * Set the auth handler on the MQTT endpoint. This handler is called when a AUTH message is
+   * received by the remote MQTT client
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/MqttEndpoint.java:343]: please leave a blank line right after the description

File: src/main/java/examples/VertxMqttServerExamples.java [hunk 4]
```diff
@@ -325,4 +327,16 @@ public void example13(Vertx vertx) {
         }
       });
   }
+
+  /**
+   * Example for authentication exchange
```
Comment (ppatierno) [inline src/main/java/examples/VertxMqttServerExamples.java:332]: please leave a blank line right after the description

File: src/main/java/io/vertx/mqtt/MqttEndpoint.java [hunk 5]
```diff
@@ -575,6 +585,15 @@ public interface MqttEndpoint {
   @GenIgnore(GenIgnore.PERMITTED_TYPE)
   Future<Integer> publish(String topic, Buffer payload, MqttQoS qosLevel, boolean isDup, boolean isRetain, int messageId, MqttProperties properties);
 
+
+  /**
+   * Sends the AUTH message to the remote MQTT client
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/MqttEndpoint.java:590]: please leave a blank line right after the description

File: src/main/java/io/vertx/mqtt/messages/codes/MqttAuthenticateReasonCode.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package io.vertx.mqtt.messages.codes;
+
+public enum MqttAuthenticateReasonCode implements MqttReasonCode {
+
+  SUCCESS((byte) 0x0),
+
+  CONTINUE_AUTHENTICATION((byte) 0x18),
+
+  RE_AUTHENTICATE((byte) 0x19),
+  ;
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/messages/codes/MqttAuthenticateReasonCode.java:10]: should this be just on the previous line?

File: src/test/java/io/vertx/mqtt/test/server/Mqtt5ServerAuthTest.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.lang.reflect.Field;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+
+/**
+ * MQTT server testing about server publish
```
Comment (ppatierno) [inline src/test/java/io/vertx/mqtt/test/server/Mqtt5ServerAuthTest.java:41]: it's not testing server publish, but server auth.

**Review by ppatierno — CHANGES_REQUESTED** [review]

**Comment by qianwj** [comment]
done.

**Review by ppatierno — APPROVED** [review]
LGTM

**Comment by ppatierno** [comment]
@vietj is there anything on your side here?

**Comment by vietj** [comment]
thank you @qianwj

**Pull request merged**